### ISSUE TITLE
Proper handling of subprocess arguments

### ISSUE THREAD

**Issue by jussiks** [JYU-IBA/potku#19]
The way Potku builds commands that are used to call external programs via the subprocess module needs to be revised. Commands are often built using basic string operations which poses a security threat when some of the arguments come from user input or file system [[0]](https://owasp.org/www-community/attacks/Command_Injection).

Arguments to subprocess should preferably be tuples or lists [[1].](https://docs.python.org/3/library/subprocess.html#frequently-used-arguments)  [shlex](https://docs.python.org/3.6/library/shlex.html) module can also used to escape potentially harmful inputs.

Edit: moving to [ctypes](https://docs.python.org/3.6/library/ctypes.html) module would be the nicest option, but that requires rewriting the C programs so that they can be used as dlls/shared libraries.

**Comment by jussiks** [comment]
`coinc` got an overhaul in #82. `get_espe` could be next. Instead using `shlex.split` like this:

https://github.com/JYU-IBA/potku/blob/e6b630ce5fda3b1bbcd61fe62cf008dc1b8a199b/modules/get_espe.py#L177-L198

`params` variable should be split "manually":

```python
params = (
    "-beam", self.__beam.ion.get_prefix(), 
    "-energy", str(self.__beam.energy}),
     [...] 
    "-ch", str(self.__channel_width),
)
```

The problem with `shlex.split` is that it just splits strings by whitespace so a file path like `/tmp/foo.bar && rm ~` becomes `["/tmp/foo.bar", "&&", "rm", "~"]`. When that list is given to subprocess.run, suddenly your home folder is gone.

Might as well self-assign this as I'm the one who started using `shlex.split`...

**Comment by jussiks** [comment]
> 
> The problem with `shlex.split` is that it just splits strings by whitespace so a file path like `/tmp/foo.bar && rm ~` becomes `["/tmp/foo.bar", "&&", "rm", "~"]`. When that list is given to subprocess.run, suddenly your home folder is gone.
>

This turned out to be wrong. `shlex.split` in conjunction with `subprocess.run` is safe and `rm -rf ~` would not get executed in the above example. The problem was that `coinc` accepts an arbitrary number of output files, which it also promptly creates after being called. That is why this test failed initially:

https://github.com/JYU-IBA/potku/blob/e6b630ce5fda3b1bbcd61fe62cf008dc1b8a199b/tests/unit/test_general_functions.py#L457-L468

The file existed on line 468 because `coinc` created it, not `touch` as I had assumed (when this test was failing, `coinc` would also create files `&&` and `touch` in the working directory).

So in conclusion, `shlex.split` can be used, but if you are dealing with arguments that can contain whitespace (for example filenames), it is better to split the arguments 'manually'.

**Comment by jussiks** [comment]
#83 refactors `get_espe` module, but no security holes were found. Potku seems to handle its command line arguments pretty well by now, but I'll go through all subprocess calls once more before closing this one.

On a related note, I added a `subprocess_utils` module in 225b85e to work with subprocesses in a more reusable manner. I'll start integrating that module with other modules once I've checked the PR on both Linux and Windows.

**Comment by jussiks** [comment]
As far as I can tell, there are no longer problematic subprocess calls. Of course in the future more calls may be added and existing ones need to be be modified, so here are some guidelines on what to do:
- Never use `shell=True`. Only run a single shell command on each subprocess call.
- Multiple processes can be pipelined by using `stdout` and `stdin` parameters. Example:
https://github.com/JYU-IBA/potku/blob/0ec87e44f6812f250c054ca6e8cdf0b8bd4aed16/modules/depth_files.py#L101-L103
- Supply shell commands as lists or a tuples where each argument is its own element.
- Prefer splitting arguments into lists or tuples manually instead of using `shlex.split`. `shlex.split` splits strings by whitespace which can cause problems with file paths etc. (although `shlex.quote` may help).
